### DISCUSSION TITLE
FastAPI seems much slower that jupyter by vscode when useing pymongo

### DISCUSSION THREAD

**Discussion by MintBlue** [fastapi/fastapi#6118] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@appmat.post("/mt")
def mtest(num: int):
    s = time.perf_counter()
    client = pymongo.MongoClient("mongodb://localhost:27017/")
    db = client['tdm']
    mycollec2 = db['ttt']
    test2Data = []
    for i in range(num):
        test2Data.append({
            "time": i,
            "a": random.randint(1, 8),
            "b": random.randint(0, 1)
        })

    mycollec2.insert_many(test2Data)

    e = time.perf_counter()

    return str(round(sys.getsizeof(test2Data) / 1024 / 1024, 2)) + "MB", e - s
```

### Description

Thanks for help.
I am a learner in Python and FastAPI.My work is very simple,using FastAPI to insert data to mongodb by pymongo.
- pymongo version : `4.2.0`.
- mongodb version:`MongoDB 4.2.8 Community Standalone`
- FastAPI Start Command:`uvicorn.run('main:app', host='127.0.0.1', port=58000, reload=True)`
- CPU:`AMD 5800H`
- RAM:`32G`

My problem is that I found inserting data using the way I mentioned above was slow ,but when I copy the code and run in vscode using jupyter directly,it's much faster than before
You can see the detail in the below table:

| num      | fastapi             | jupyter              |
| -------- | ------------------- | -------------------- |
| 1000     | 0.06569519999902695 | 0.017709599998852354 |
| 10000    | 0.23583729999882053 | 0.10930559999906109  |
| 100000   | 2.2969444999998814  | 1.0585518999996566   |
| 1000000  | 23.430175199999212  | 9.922307700000601    |
| 10000000 | 238.93187810000018  | 98.82192099999884    |

So did I do something wrong or how could I make FastAPI running faster when inserting data in my project?

### Operating System

Windows

### Operating System Details

Window11 22622.598

### FastAPI Version

0.85.0

### Python Version

Python 3.10.7

**Comment by iudeen** [comment, marked as answer]
By Jupyter, you mean you have direct functions instead of a FastAPI app?

If that's the case, do the following:

Move the connection outside your route function. 

```python
client = pymongo.MongoClient("mongodb://localhost:27017/")

@app.route("/")
def foo():
   ...
```

This will help overcome connection overhead on multiple requests.

~~I am also guessing the slowness can be because of how FastAPI executes the function.~~ That is it runs using a threadpool for sync routes. I am not sure if Pymongo uses multi-thread for insert many...

**Comment by MintBlue** [comment]
> By Jupyter, you mean you have direct functions instead of a FastAPI app?
> 
> If that's the case, do the following:
> 
> Move the connection outside your route function.
> 
> ```python
> client = pymongo.MongoClient("mongodb://localhost:27017/")
> 
> @app.route("/")
> def foo():
>    ...
> ```
> 
> This will help overcome connection overhead on multiple requests.
> 
> I am also guessing the slowness can be because of how FastAPI executes the function. That is it runs using a threadpool for sync routes. I am not sure if Pymongo uses multi-thread for insert many...

Thanks for your reply.
I direct  run the inserting data code instead of a FastAPI app.
I tried to move the connection outside my route function,but the time cost to insert data seems to have no change.

**Comment by notypecheck** [comment]
I think in this case there should be no difference between running code in a fastapi endpoint and running the code as-is 🤔

**Comment by iudeen** [comment]
@ThirVondukr you are right!

Did some tests, I don't see any difference. And Pymongo does not use multi-threading when we call `insert_many()` rather, issues a single command to MongoDB. 

Here are the results:
```
100
-------------------------------
FastaAPI ["0.0MB",0.02326849999371916]
Pure Python ('0.0MB', 0.027795600006356835)

1000
-------------------------------
FastAPI ["0.01MB",0.13195249997079372]
Pure Python ('0.01MB', 0.12988080002833158)

10000
-------------------------------
FastAPI ["0.08MB",0.537144000059925]
Pure Python ('0.08MB', 0.6174862999469042)

100000
-------------------------------
FastAPI ["0.76MB",3.919053899939172]
Pure Python ('0.76MB', 4.565466900006868)
```

**Comment by MintBlue** [comment]
Sorry for taking everyone's time,it's my mistake.I run the code in debugger mode in pycharm.
I have not realized that the debugger mode would cause performance lost so much before when using python.
And when I running the exe packaged by pyinstall,the cost time is close to jupyter.

**Thanks again**

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue @MintBlue  👍
